### ISSUE TITLE
QA: ENV variables while deploying Node apps with Dockerfile

### ISSUE THREAD

**Issue by adamtrip** [prom3theu5/aspirational-manifests#75]
I've been experimenting a lot with deploying ASP.NET API's with NextJS frontend with the new tag: `AsDockerfileInManifest()`.
While running the app via Visual Studio, all works like a charm, env variables are injected and all works.
However when deploying a sample app, none of that variables are passed.

I'm going to provide my code so you can get a better view:

```
using Microsoft.Extensions.Hosting;

var builder = DistributedApplication.CreateBuilder(args);

var webapi = builder.AddProject<Projects.nexttest_WebAPI>("aspirate-api");

if (builder.Environment.IsProduction())
{
    builder.AddNpmApp("aspirate-nextjs", "../nextjs-aspire-demo", "prod")
        .WithReference(webapi)
        .WithServiceBinding(3000, scheme: "http", env: "PORT")
        .AsDockerfileInManifest();
}
else
{
    builder.AddNpmApp("aspirate-nextjs", "../nextjs-aspire-demo", "dev")
        .WithReference(webapi)
        .WithServiceBinding(3000, scheme: "http", env: "PORT")
        .AsDockerfileInManifest();
}


builder.Build().Run();
```

Dockerfile:
```
FROM node:18-alpine AS base

# Install dependencies only when needed
FROM base AS deps
# Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
RUN apk add --no-cache libc6-compat
WORKDIR /app

# Install dependencies based on the preferred package manager
COPY package.json yarn.lock* package-lock.json* pnpm-lock.yaml* ./
RUN \
  if [ -f yarn.lock ]; then yarn --frozen-lockfile; \
  elif [ -f package-lock.json ]; then npm ci; \
  elif [ -f pnpm-lock.yaml ]; then yarn global add pnpm && pnpm i --frozen-lockfile; \
  else echo "Lockfile not found." && exit 1; \
  fi


# Rebuild the source code only when needed
FROM base AS builder
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY . .

# Next.js collects completely anonymous telemetry data about general usage.
# Learn more here: https://nextjs.org/telemetry
# Uncomment the following line in case you want to disable telemetry during the build.
# ENV NEXT_TELEMETRY_DISABLED 1

RUN npm run build

# If using npm comment out above and use below instead
# RUN npm run build

# Production image, copy all the files and run next
FROM base AS runner
WORKDIR /app

ENV NODE_ENV production
# Uncomment the following line in case you want to disable telemetry during runtime.
# ENV NEXT_TELEMETRY_DISABLED 1

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

COPY --from=builder /app/public ./public

# Set the correct permission for prerender cache
RUN mkdir .next
RUN chown nextjs:nodejs .next

# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

USER nextjs

EXPOSE 3000

ENV PORT 3000

# set hostname to localhost
ENV HOSTNAME "0.0.0.0"

# server.js is created by next build from the standalone output
# https://nextjs.org/docs/pages/api-reference/next-config-js/output
CMD ["node", "server.js"]
```

I believe that the reason why the variables are not being passed is because they are not being defined on Dockerfile and that raises my question:
Since having a dockerfile is required to deploy, how do pass those variables?

Thanks in advance and any other info you need, feel free to ask

**Comment by prom3theu5** [comment]
Hi

Good question really - this is something aspire needs to be expanded to incorporate really. I don't know if they have it on the horizon, but they should really support the concept of build-args.

The extension method `.AsDockerfileInManifest();` simply outputs the npm app as a dockerfile when writing the manifest - it doesn't actually run it as a container when debugging in VS

Because the manifest only contains env vars - if we want to support this now one way to do this would be to utilise env vars that the container can have in the manifest

Then in your dockerfile you'd have to have ARGs, and set the ENV's from the build args passed in

Like so

```dockerfile
# expect a build-time variable
ARG A_VARIABLE
# use the value to set the ENV var default
ENV an_env_var=$A_VARIABLE
# if not overridden, that value of an_env_var will be available to your containers!
```

Utilising this method - I could change it so that all env vars from the manifest file on a dockerfile component, also get passed as `--build-arg` values during container building?

Would this be acceptable?

**Comment by adamtrip** [comment]
As a temporary solution, I think it will suffice.
Maybe I'll open an issue on aspire questioning this and link this post to see if they have anything in mind.

**Comment by prom3theu5** [comment]
Sure thing
Even if aspire is updated to include separate build args in the manifest - you will still have to add the args to your Dockerfile like above.

I've just released version `0.1.17-preview` - this includes the support for passing env vars as build args
